File: frontend/src/lib/hours.ts
/**
 * Open-now / next-open helpers for ServiceLocation.hours dicts.
 * Mirrors backend/app/services/hours.py — keep the two parsers in sync.
 */

import type { ServiceLocation } from "@/types";

const DAYS = [
  "monday",
  "tuesday",
  "wednesday",
  "thursday",
  "friday",
  "saturday",
  "sunday",
] as const;
type Day = (typeof DAYS)[number];

const DAY_LABEL: Record<Day, string> = {
  monday: "Monday",
  tuesday: "Tuesday",
  wednesday: "Wednesday",
  thursday: "Thursday",
  friday: "Friday",
  saturday: "Saturday",
  sunday: "Sunday",
};

const RANGE_SPLIT_RE = /\s*[-–—]\s*/;
const TIME_RE = /^\s*(\d{1,2})(?::(\d{2}))?\s*([AaPp])\.?\s*[Mm]\.?\s*$/;

interface ParsedTime {
  hour: number;
  minute: number;
}

type ParsedRange =
  | { kind: "range"; open: ParsedTime; close: ParsedTime }
  | { kind: "twentyFour" }
  | { kind: "closed" };

function parseTime(token: string): ParsedTime | null {
  const m = TIME_RE.exec(token);
  if (!m) return null;
  let hour = parseInt(m[1], 10);
  const minute = m[2] ? parseInt(m[2], 10) : 0;
  const meridiem = m[3].toLowerCase();
  if (meridiem === "p" && hour !== 12) hour += 12;
  else if (meridiem === "a" && hour === 12) hour = 0;
  if (hour < 0 || hour > 23 || minute < 0 || minute > 59) return null;
  return { hour, minute };
}

function parseRange(label: string | undefined): ParsedRange {
  const text = (label ?? "").trim().toLowerCase();
  if (
    !text ||
    text === "closed" ||
    text === "n/a" ||
    text === "by appointment" ||
    text === "appointment only"
  ) {
    return { kind: "closed" };
  }
  if (text.includes("24") && text.includes("hour")) {
    return { kind: "twentyFour" };
  }
  const parts = (label ?? "").trim().split(RANGE_SPLIT_RE);
  if (parts.length !== 2) return { kind: "closed" };
  const open = parseTime(parts[0]);
  const close = parseTime(parts[1]);
  if (!open || !close) return { kind: "closed" };
  return { kind: "range", open, close };
}

function dayKey(date: Date): Day {
  // JS getDay: Sunday=0..Saturday=6 → map to monday-first DAYS array
  return DAYS[(date.getDay() + 6) % 7];
}

function minutesSinceMidnight(date: Date): number {
  return date.getHours() * 60 + date.getMinutes();
}

function rangeMinutes(t: ParsedTime): number {
  return t.hour * 60 + t.minute;
}

export function isOpenNow(
  location: Pick<ServiceLocation, "hours">,
  now: Date = new Date(),
): boolean {
  if (!location.hours) return false;
  const today = dayKey(now);
  const parsed = parseRange(location.hours[today]);
  if (parsed.kind === "closed") return false;
  if (parsed.kind === "twentyFour") return true;
  const cur = minutesSinceMidnight(now);
  const open = rangeMinutes(parsed.open);
  const close = rangeMinutes(parsed.close);
  if (close <= open) return cur >= open || cur < close;
  return cur >= open && cur < close;
}

function formatTime(t: ParsedTime): string {
  const hour12 = t.hour % 12 || 12;
  const suffix = t.hour < 12 ? "am" : "pm";
  if (t.minute === 0) return `${hour12}${suffix}`;
  return `${hour12}:${t.minute.toString().padStart(2, "0")}${suffix}`;
}

export function nextOpenLabel(
  location: Pick<ServiceLocation, "hours">,
  now: Date = new Date(),
): string | null {
  if (!location.hours || isOpenNow(location, now)) return null;
  for (let offset = 0; offset < 8; offset++) {
    const check = new Date(now.getTime() + offset * 86_400_000);
    const day = dayKey(check);
    const parsed = parseRange(location.hours[day]);
    if (parsed.kind === "closed") continue;
    if (parsed.kind === "twentyFour") {
      return offset > 0 ? "Opens at midnight" : "Open 24 hours";
    }
    if (
      offset === 0 &&
      minutesSinceMidnight(check) >= rangeMinutes(parsed.open)
    )
      continue;
    const opensAt = formatTime(parsed.open);
    if (offset === 0) return `Opens ${opensAt}`;
    if (offset === 1) return `Opens tomorrow ${opensAt}`;
    return `Opens ${DAY_LABEL[day]} ${opensAt}`;
  }
  return null;
}

/** Returns true when ANY of the service's locations is currently open. */
export function serviceOpenNow(
  locations: Pick<ServiceLocation, "hours">[],
  now: Date = new Date(),
): boolean {
  return locations.some((loc) => isOpenNow(loc, now));
}

/** Compact label like "Open now" / "Closed — opens tomorrow 8am". */
export function openStatusLabel(
  location: Pick<ServiceLocation, "hours">,
  now: Date = new Date(),
): { open: boolean; label: string } {
  if (isOpenNow(location, now)) return { open: true, label: "Open now" };
  const next = nextOpenLabel(location, now);
  return { open: false, label: next ? `Closed · ${next}` : "Hours not listed" };
}

File: backend/app/services/hours.py
"""Open-now / next-open helpers for ServiceLocation.hours dicts.

Hours dicts in seed data look like:
    {"monday": "8:00 AM – 5:00 PM", ..., "saturday": "Closed"}
or sparse:
    {"saturday": "9:00 AM – 11:00 AM"}
or:
    {"monday": "24 hours", ...}

Times accept hyphen, en-dash, or em-dash separators.  Days that are
absent from the dict are treated as Closed.
"""

from __future__ import annotations

import re
from datetime import datetime, time, timedelta
from zoneinfo import ZoneInfo

from app.models import ServiceLocation

DEFAULT_TZ = "America/Chicago"

_DAYS = ["monday", "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday"]
_DAY_LABEL = {d: d.capitalize() for d in _DAYS}

_RANGE_SPLIT_RE = re.compile(r"\s*[-–—]\s*")
_TIME_RE = re.compile(r"^\s*(\d{1,2})(?::(\d{2}))?\s*([AaPp])\.?\s*[Mm]\.?\s*$")


def _parse_time(token: str) -> time | None:
    m = _TIME_RE.match(token)
    if not m:
        return None
    hour = int(m.group(1))
    minute = int(m.group(2) or 0)
    meridiem = m.group(3).lower()
    if meridiem == "p" and hour != 12:
        hour += 12
    elif meridiem == "a" and hour == 12:
        hour = 0
    if not (0 <= hour <= 23 and 0 <= minute <= 59):
        return None
    return time(hour=hour, minute=minute)


def _parse_range(label: str) -> tuple[time, time] | None | str:
    """Return (open, close) tuple, ``None`` for closed, or the literal
    string ``"24"`` for round-the-clock."""
    text = (label or "").strip().lower()
    if not text or text in {"closed", "n/a", "by appointment", "appointment only"}:
        return None
    if "24" in text and "hour" in text:
        return "24"
    parts = _RANGE_SPLIT_RE.split(label.strip(), maxsplit=1)
    if len(parts) != 2:
        return None
    start = _parse_time(parts[0])
    end = _parse_time(parts[1])
    if start is None or end is None:
        return None
    return (start, end)


def _now(tz: str | None) -> datetime:
    return datetime.now(ZoneInfo(tz or DEFAULT_TZ))


def is_open_now(
    location: ServiceLocation,
    *,
    now: datetime | None = None,
    tz: str = DEFAULT_TZ,
) -> bool:
    """Return ``True`` if the location is currently open."""
    if not location.hours:
        return False
    current = now or _now(tz)
    today = _DAYS[current.weekday()]
    parsed = _parse_range(location.hours.get(today, ""))
    if parsed is None:
        return False
    if parsed == "24":
        return True
    open_t, close_t = parsed
    now_t = current.timetz().replace(tzinfo=None)
    if close_t <= open_t:
        # Overnight range (e.g. 22:00–02:00) — open if now ≥ open or now < close
        return now_t >= open_t or now_t < close_t
    return open_t <= now_t < close_t


def next_open_label(
    location: ServiceLocation,
    *,
    now: datetime | None = None,
    tz: str = DEFAULT_TZ,
) -> str | None:
    """Return a short human label for the next opening, or ``None`` when
    the location is open now or has no schedulable hours."""
    if not location.hours or is_open_now(location, now=now, tz=tz):
        return None
    current = now or _now(tz)
    for offset in range(0, 8):
        check = current + timedelta(days=offset)
        day = _DAYS[check.weekday()]
        parsed = _parse_range(location.hours.get(day, ""))
        if parsed is None:
            continue
        if parsed == "24":
            return "Opens at midnight" if offset > 0 else "Open 24 hours"
        open_t, _ = parsed
        if offset == 0 and check.timetz().replace(tzinfo=None) >= open_t:
            continue
        opens_at = _format_time(open_t)
        if offset == 0:
            return f"Opens {opens_at}"
        if offset == 1:
            return f"Opens tomorrow {opens_at}"
        return f"Opens {_DAY_LABEL[day]} {opens_at}"
    return None


def _format_time(t: time) -> str:
    hour12 = t.hour % 12 or 12
    suffix = "am" if t.hour < 12 else "pm"
    if t.minute == 0:
        return f"{hour12}{suffix}"
    return f"{hour12}:{t.minute:02d}{suffix}"


def service_open_now(service_locations: list[ServiceLocation], **kwargs) -> bool:
    """Return ``True`` when *any* of the service's locations are open."""
    return any(is_open_now(loc, **kwargs) for loc in service_locations)
